samcli/commands/publish/command.py

```diff
@@ -31,31 +33,40 @@
 """
 SHORT_HELP = "Publish a packaged AWS SAM template to the AWS Serverless Application Repository."
 SERVERLESSREPO_CONSOLE_URL = "https://console.aws.amazon.com/serverlessrepo/home?region={}#/published-applications/{}"
+SEMANTIC_VERSION_HELP = "Optional. The value provided here overwrites SemanticVersion in the template metadata."
+SEMANTIC_VERSION = 'SemanticVersion'
 
 
 @click.command("publish", help=HELP_TEXT, short_help=SHORT_HELP)
 @template_common_option
+@click.option('--semantic-version', help=SEMANTIC_VERSION_HELP)
 @aws_creds_options
 @cli_framework_options
 @pass_context
-def cli(ctx, template):
+def cli(ctx, template, semantic_version):
     # All logic must be implemented in the ``do_cli`` method. This helps with easy unit testing
 
-    do_cli(ctx, template)  # pragma: no cover
+    do_cli(ctx, template, semantic_version)  # pragma: no cover
 
 
-def do_cli(ctx, template):
+def do_cli(ctx, template, semantic_version):
     """Publish the application based on command line inputs."""
     try:
         template_data = get_template_data(template)
     except ValueError as ex:
         click.secho("Publish Failed", fg='red')
         raise UserException(str(ex))
 
+    # Overwrite SemanticVersion in template metadata when provided in command input
+    if semantic_version and SERVERLESS_REPO_APPLICATION in template_data.get(METADATA, {}):
+        template_data.get(METADATA).get(SERVERLESS_REPO_APPLICATION)[SEMANTIC_VERSION] = semantic_version
+        with open(template, 'w') as fp:
```

Comment:
What is the expectation here? What happens when this is a template that is generated and isn't the exact `source` copy customers edit themselves?

Reply:
The expectation is that the template used to publish will always be up to date, and the latest semantic version can be tracked in the template. In most cases it's not the source copy since the template needs to be packaged first, but modifying the source copy requires additional inputs (like path to the source template), even with that we can't ensure it's the "source". I think it's better to keep the command simple and just update the published template.

Reply:
I am not crazy about writing this back to the template. Till now, none of the commands writes or updates the template that is being handled. Can we pass the semantic version to the `publish_application` instead? Customers using `build` will NEED to update the source template anyways, as running build will squash this change to the template. I would prefer a solution that matches the expectation of editing the source rather than possibly editing the generated sources.

Reply:
Made the change.